odl/trafos/backends/pywt_bindings.py

```diff
@@ -0,0 +1,523 @@
+                         'level is {}'
+                         ''.format(nscales, max_level))
+
+    # dwt_coeff_len calculates the number of coefficients at the next
+    # scaling level given the input shape, the length of the filter and
+    # the applied mode.
+    # We use this in the following way (per dimension):
+    # - length[0] = original data length
+    # - length[n+1] = dwt_coeff_len(length[n], ...)
+    # - until n = nscales
+    shape_list = [shape]
+    for scale in range(nscales):
+        shp = tuple(pywt.dwt_coeff_len(n, filter_len=wbasis.dec_len,
+                                       mode=pywt_mode)
+                    for n in shape_list[scale])
+        shape_list.append(shp)
+
+    # Add a duplicate of the last entry for the approximation coefficients
+    shape_list.append(shape_list[-1])
+
+    # We created the list in reversed order compared to what pywt expects
+    shape_list.reverse()
+    return shape_list
+
+
+def pywt_coeff_to_array(coeff, shape_list):
+    """Convert a Pywavelets coefficient list into a flat array.
+
+    Related to 1D, 2D and 3D multilevel discrete wavelet transforms.
+
+    Parameters
+    ----------
+    coeff : ordered `sequence`
+        Sequence of coefficients organized in the following way:
+
+        In 1D:
+
+        ``[aN, (dN), ..., (d1)]``
+
+        In 2D:
+
+        ``[aaN, (adN, daN, ddN), ..., (ad1, da1, dd1)]``
+
+        In 3D:
+
+        ``[aaaN, (aadN, adaN, addN, daaN, dadN, ddaN, dddN), ...
+        (aad1, ada1, add1, daa1, dad1, dda1, ddd1)]``
+
+        The abbreviations are "a" for "approximation" and "d" for "detail".
+        The letter position encodes the coefficient type in the respective
+        axis. ``N`` stands for the number of scaling levels.
+
+    shape_list : `sequence`
+        Sequence containing the shapes of the wavelet (approximation
+        and detail) coefficients at different scaling levels.
+
+        ``shape_list[0]`` = shape of approximation coefficients at
+        the coarsest level,
+
+        ``shape_list[1]`` = shape of the detailed coefficients at
+        the coarsest level,
+
+        ``shape_list[N]`` = shape of the detailed coefficients at
+        the finest level,
+
+        ``shape_list[N+1]`` = shape of original image,
+
+        ``N`` =  number of scaling levels
+
+    Returns
+    -------
+    arr : `numpy.ndarray`
+        Flattened and concatenated coefficient array.
+        The length of the array depends on the shape of the input image to
+        be transformed and on the chosen wavelet basis.
+      """
+    flat_sizes = [np.prod(shp) for shp in shape_list[:-1]]
+    ndim = len(shape_list[0])
+    dcoeffs_per_scale = 2 ** ndim - 1
+
+    flat_total_size = flat_sizes[0] + dcoeffs_per_scale * sum(flat_sizes[1:])
+    flat_coeff = np.empty(flat_total_size)
+
+    start = 0
+    stop = flat_sizes[0]
+    flat_coeff[start:stop] = coeff[0].ravel()
```

Comment:
We implicitly assume its members to be arrays. If we expect at any point to get array-like members,  we need to use `np.ravel` or `np.asarray(x).ravel()`.